azkaban-common/src/main/java/azkaban/project/ProjectManager.java

```diff
@@ -86,12 +98,115 @@ public ProjectManager(ProjectLoader loader, Props props) {
     new XmlValidatorManager(prop);
     loadAllProjects();
     loadProjectWhiteList();
+
+    CleanerThread cleanerThread = new CleanerThread();
+    cleanerThread.start();
+  }
+
+  /**
+   * Cleaner thread to clean projects. Runs every day.
+  **/
+  private class CleanerThread extends Thread {
+
+    // runs every day
+    private static final long CLEANER_THREAD_WAIT_INTERVAL_MS = 24 * 60 * 60 * 1000;
+    private static final String DELETER = "azkaban";
+
+    private boolean shutdown = false;
+    private long lastProjectFileCleanTime = -1;
+    private long lastProjectCleanTime = -1;
+
+    public CleanerThread() {
+      this.setName("ProjectManager-Cleaner-Thread");
+      logger.info("Starting cleaner thread: " + this.getName());
+    }
+
+    @SuppressWarnings("unused")
+    public void shutdown() {
+      shutdown = true;
+      this.interrupt();
+    }
+
+    public void run() {
+      while (!shutdown) {
+        synchronized (this) {
```

Comment:
Did we use sync only for the wait() below? Can't we rather use sleep(ms) - this won't need sync ?
My issue or doubt here is, the functionality being done is a bit heavy to be done under a single lock. Either remove sync or keep only necessary critical section with sync

Reply:
ack, removed sync and used sleep, this is for my context a single-threaded processing, since we are initializing only in ProjectLoader